### PR TITLE
Implement ranges::replace_copy_if

### PR DESCRIPTION
**Pull request by miscco**
This implements ranges::replace_copy_if

There is still the compile error wrt `indirect_copyable` same as for `replace_copy`

### CODE REVIEW

File: stl/inc/algorithm [hunk 1]
```diff
@@ -3095,6 +3095,67 @@ _FwdIt2 replace_copy_if(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Last, _FwdIt2 _Dest, _
...
+                _STD move(_Result), _Pass_fn(_Pred), _Newval, _Pass_fn(_Proj));
+
+            _Seek_wrapped(_First, _STD move(_UResult.in));
+            return {_STD move(_First), _STD move(_UResult.out)};
+        }
+        // clang-format on
+    private:
+        template <class _It, class _Se, class _Ty, class _Out, class _Pj, class _Pr>
+        _NODISCARD constexpr replace_copy_if_result<_It, _Out> _Replace_copy_if_unchecked(
+            _It _First, const _Se _Last, _Out _Result, _Pr _Pred, const _Ty& _Newval, _Pj _Proj) const {
```
Comment (CaseyCarter) [inline stl/inc/algorithm:3137]:
```suggestion
        _NODISCARD static constexpr replace_copy_if_result<_It, _Out> _Replace_copy_if_unchecked(
            _It _First, const _Se _Last, _Out _Result, _Pr _Pred, const _Ty& _Newval, _Pj _Proj) {
```

File: tests/std/tests/P0896R4_ranges_alg_replace_copy_if/test.cpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+STATIC_ASSERT(same_as<decltype(ranges::replace_copy(borrowed<false>{}, static_cast<int*>(nullptr), matches, 5)),
+    ranges::replace_copy_if_result<ranges::dangling, int*>>);
+STATIC_ASSERT(same_as<decltype(ranges::replace_copy(borrowed<true>{}, static_cast<int*>(nullptr), matches, 5)),
+    ranges::replace_copy_if_result<int*, int*>>);
+
+struct instantiator {
+    static constexpr P input[5]    = {{0, 99}, {1, 47}, {2, 99}, {3, 47}, {4, 99}};
+    static constexpr P expected[5] = {{0, 99}, {47, 1}, {2, 99}, {47, 1}, {4, 99}};
+
+    template <ranges::input_range Read>
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_ranges_alg_replace_copy_if/test.cpp:29]:
```suggestion
    template <ranges::input_range Read, ranges::indirectly_writable<ranges::range_reference_t<Read>> Write>
```

File: tests/std/tests/P0896R4_ranges_alg_replace_copy_if/test.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    static constexpr void call() {
+        using ranges::replace_copy_if;
+        { // Validate iterator + sentinel overload
+            Read wrapped_input{input};
+
+            auto result = replace_copy_if(
+                wrapped_input.begin(), wrapped_input.end(), Write{output}, matches, P{47, 1}, get_second);
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_ranges_alg_replace_copy_if/test.cpp:36]: `output` isn't declared. (And on 45.)

File: tests/std/tests/P0896R4_ranges_alg_replace_copy_if/test.cpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            STATIC_ASSERT(same_as<decltype(result), replace_copy_if_result<iterator_t<Read>, Write>>);
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_ranges_alg_replace_copy_if/test.cpp:37]:
```suggestion
            STATIC_ASSERT(same_as<decltype(result), replace_copy_if_result<ranges::iterator_t<Read>, Write>>);

```
(And on 46.)

Comment (miscco) [reply tests/std/tests/P0896R4_ranges_alg_replace_copy_if/test.cpp:37]: I changed it to using declarations as in the other tests

File: tests/std/tests/P0896R4_ranges_alg_replace_copy_if/test.cpp (hunk 3 above)
Comment (miscco) [reply tests/std/tests/P0896R4_ranges_alg_replace_copy_if/test.cpp:36]: Done

File: tests/std/tests/P0896R4_ranges_alg_replace_copy_if/test.cpp (hunk 2 above)
Comment (miscco) [reply tests/std/tests/P0896R4_ranges_alg_replace_copy_if/test.cpp:29]: Done

File: stl/inc/algorithm (hunk 1 above)
Comment (miscco) [reply stl/inc/algorithm:3137]: Done

File: stl/inc/algorithm [hunk 5]
```diff
@@ -3095,6 +3095,67 @@ _FwdIt2 replace_copy_if(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Last, _FwdIt2 _Dest, _
...
+            _It _First, _Se _Last, _Out _Result, _Pr _Pred, const _Ty& _Newval, _Pj _Proj = {}) const {
+            _Adl_verify_range(_First, _Last);
+            auto _UResult = _Replace_copy_if_unchecked(_Get_unwrapped(_STD move(_First)),
+                _Get_unwrapped(_STD move(_Last)), _STD move(_Result), _Pass_fn(_Pred), _Newval, _Pass_fn(_Proj));
+
+            _Seek_wrapped(_First, _STD move(_UResult.in));
+            return {_STD move(_First), _STD move(_UResult.out)};
+        }
+
+        template <input_range _Rng, class _Ty, output_iterator<const _Ty2&> _Out, class _Pj = identity,
```
[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/algorithm:3121]: I even found another `_Ty1`

File: stl/inc/algorithm [hunk 6]
```diff
@@ -3095,6 +3095,67 @@ _FwdIt2 replace_copy_if(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Last, _FwdIt2 _Dest, _
 
 #ifdef __cpp_lib_concepts
 namespace ranges {
+    // ALIAS TEMPLATE replace_copy_result
```
[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/algorithm:3098]: Copy paste...

File: stl/inc/algorithm [hunk 7]
```diff
@@ -3095,6 +3095,67 @@ _FwdIt2 replace_copy_if(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Last, _FwdIt2 _Dest, _
...
+            _Adl_verify_range(_First, _Last);
+            auto _UResult = _Replace_copy_if_unchecked(_Get_unwrapped(_STD move(_First)),
+                _Get_unwrapped(_STD move(_Last)), _STD move(_Result), _Pass_fn(_Pred), _Newval, _Pass_fn(_Proj));
+
+            _Seek_wrapped(_First, _STD move(_UResult.in));
+            return {_STD move(_First), _STD move(_UResult.out)};
+        }
+
+        template <input_range _Rng, class _Ty, output_iterator<const _Ty2&> _Out, class _Pj = identity,
+            indirect_unary_predicate<projected<_It, _Pj>> _Pr>
```
[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/algorithm:3122]: Done

**Comment by miscco** [comment]
Narf, I should really take thetime to read everything a second Time if I do not expect it to compileanyway. Sorry for the noise

File: stl/inc/algorithm [hunk 8]
```diff
@@ -3095,6 +3095,67 @@ _FwdIt2 replace_copy_if(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Last, _FwdIt2 _Dest, _
...
+
+            _Seek_wrapped(_First, _STD move(_UResult.in));
+            return {_STD move(_First), _STD move(_UResult.out)};
+        }
+        // clang-format on
+    private:
+        template <class _It, class _Se, class _Ty, class _Out, class _Pj, class _Pr>
+        _NODISCARD static constexpr replace_copy_if_result<_It, _Out> _Replace_copy_if_unchecked(
+            _It _First, const _Se _Last, _Out _Result, _Pr _Pred, const _Ty& _Newval, _Pj _Proj) {
+            // copy [_First, _Last) to _Out while replacing _Oldval with _Newval if projected _Oldval fulfuills _Pred
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:3138]: Typo, and `_Out` is a type while `_Result` is a variable:

```suggestion
            // copy [_First, _Last) to _Result while replacing _Oldval with _Newval if projected _Oldval fulfills _Pred
```

**Comment by miscco** [comment]
Dropping in favor of #983

**Pull request closed without merging**